extra_data/components.py

```diff
@@ -926,6 +738,311 @@ def zip_trains_pulses(trains, pulses):
+        return MultimodKeyData(det_selected, self.key)
+
+    def select_trains(self, trains):
+        return self._with_selected_det(self.det.select_trains(trains))
+
+    def split_trains(self, parts=None, trains_per_part=None, frames_per_part=None):
+        for det_split in self.det.split_trains(parts, trains_per_part, frames_per_part):
+            yield self._with_selected_det(det_split)
+
+    def ndarray(self, *, fill_value=None, out=None, roi=(), astype=None, module_gaps=False):
+        """Get data as a plain NumPy array with no labels"""
+        train_ids = np.asarray(self.det.train_ids)
+
+        module_dim = self.det.n_modules if module_gaps else len(self.modno_to_keydata)
+
+        out_shape = ((module_dim, len(train_ids))
+                     # Shape of 1 frame for 1 module with the ROI applied:
+                     + roi_shape(self._eg_keydata.entry_shape, roi))
+
+        if out is None:
+            dtype = self._eg_keydata.dtype if astype is None else np.dtype(astype)
+            out = _out_array(out_shape, dtype, fill_value=fill_value)
+        elif out.shape != out_shape:
+            raise ValueError(f'requires output array of shape {out_shape}')
+
+        for i, (modno, kd) in enumerate(sorted(self.modno_to_keydata.items())):
+            mod_ix = (modno - self.det._modnos_start_at) if module_gaps else i
+            for chunk in kd._data_chunks:
+                for tgt_slice, chunk_slice in self.det._split_align_chunk(chunk, train_ids):
+                    chunk.dataset.read_direct(
+                        out[mod_ix, tgt_slice], source_sel=(chunk_slice,) + roi
+                    )
+        return out
+
+    def xarray(self, *, fill_value=None, roi=(), astype=None):
+        from xarray import DataArray
+        arr = self.ndarray(fill_value=fill_value, roi=roi, astype=astype)
+
+        coords = {'module': self.modules, 'trainId': self.train_id_coordinates()}
+        return DataArray(arr, dims=self.dimensions, coords=coords)
+
+    def dask_array(self, *, labelled=False, fill_value=None, astype=None):
+        from dask.delayed import delayed
+        from dask.array import concatenate, from_delayed
+
+        entry_size = (self.dtype.itemsize *
+            len(self.modno_to_keydata) * np.product(self._eg_keydata.entry_shape)
+        )
+        # Aim for 1GB chunks, with an arbitrary maximum of 256 trains
+        split = self.split_trains(frames_per_part=min(1024 ** 3 / entry_size, 256))
+
+        arr = concatenate([from_delayed(
+            delayed(c.ndarray)(fill_value=fill_value, astype=astype),
+            shape=c.shape, dtype=self.dtype
+        ) for c in split], axis=1)
+
+        if labelled:
+            from xarray import DataArray
+            coords = {'module': self.modules, 'trainId': self.train_id_coordinates()}
+            return DataArray(arr, dims=self.dimensions, coords=coords)
+
+        return arr
+
+
+class XtdfImageMultimodKeyData(MultimodKeyData):
```

Comment:
Not a comment but out of curiosity, is only `XtdfImageMultimodKeyData` fundamentally able to be pulse-resolved in the sense of having `INDEX/<source>/count > 1`?

Reply:
Yup, it's only expected for the Xtdf classes. The base `MultimodKeyData` can still handle JUNGFRAU-style data, where pulses are stored along a second axis.

Reply:
I see, thanks. I asking because of `REMI` (to be renamed `DLD`), where I already pulse-resolved data as well. At the time I didn't fully realize that is indeed limited to XTDF currently, any non-XTDF source injects additional axes. On the other hand, it's especially useful here as the number of pulses is not guaranteed to stay constant.
In the field however, it does tend to require a few additional hoops to jump through for analysis and I'd like to simplify that with some `extra_data.components`.